app/eth2wrap/valcache.go

```diff
@@ -65,28 +71,40 @@ func (c *ValidatorCache) Trim() {
 	defer c.mu.Unlock()
 
 	c.active = nil
+	c.complete = nil
 }
 
-// cached returns the cached validators and true if they are available.
-func (c *ValidatorCache) cached() (ActiveValidators, bool) {
+// activeCached returns the cached active validators and true if they are available.
+func (c *ValidatorCache) activeCached() (ActiveValidators, bool) {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 
 	return c.active, c.active != nil
 }
 
-// Get returns the cached active validators, or fetches them if not available populating the cache.
-func (c *ValidatorCache) Get(ctx context.Context) (ActiveValidators, error) {
-	if cached, ok := c.cached(); ok {
-		return cached, nil
+// cached returns the cached complete validators and true if they are available.
+func (c *ValidatorCache) cached() (CompleteValidators, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.complete, c.complete != nil
+}
+
+// Get returns the cached active validators, cached complete Validators response, or fetches them if not available populating the cache.
+func (c *ValidatorCache) Get(ctx context.Context) (ActiveValidators, CompleteValidators, error) {
+	completeCached, completeOk := c.cached()
+	activeCached, activeOk := c.activeCached()
+
+	if completeOk && activeOk {
```

Comment:
This looks like "premature optimization" to me. If this is called once per slot, then acquiring write lock vs read lock shall not really matter. As the result, the code looks more complicated than it should be.

Reply:
I see your point.

Previous maintainer added that code path, but I believe before the RWLock. Will re-check.